### PR TITLE
BLU cleanup

### PR DESCRIPTION
**Pull request by Enedin9000**
**_I affirm:_**
- [X] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [X] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [X] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

- Individual spell changes:
    - Updated TP values. A lot of spells had lower TP values for 150/300/Azure, which doesnt't make sense.
    - Updated WSC values, though most were correct.
    - Added spell ecosystem.
    - Streamlined the way spells of the same type are coded.
    - Added effect changes:
        - Consolidated all AE into one function.
        - Resistance now influences duration properly.
        - Effects that decay over time now work, such as Wild Oats (VIT down).
        - Some spells had 0 duration.
       
- Physical Blue Magic spell changes:
    - Simplified some functions.
    - Added Physical Potency merits effect.
    - Added Azure Lore effect on multiplier.
    - Damage is now influenced by monster correlation.
    - All physical spells now need to hit before the added effect (AE) can kick in.
    - All physical spells with an AE get a resistance check for the AE.
   
- Magical Blue Magic spell changes:
    - Simplified Azure Lore's multiplier code.
    - Simplified main attribute checking code (INT/MND/CHR).
    - Damage is now influenced by monster correlation.
    - Added multi-target damage reduction.
    - Added effect changes:
        - Consolidated all AE into one function.
        - Resistance now influences duration properly.
       
- Enhancing Blue Magic spell changes:
    - Consolidated Diffusion's effect on duration into one function.
    - Effects that decay over time now work, such as Saline Coat (Magic Defense Boost).
   
- Enfeebling Blue Magic spell changes:
    - Consolidated enfeebling spells into one function, where possible.
    - Added isGaze/isConal flags.
    - Put the resist threshold in line with other enfeebles (>= 0.5) and halved duration when half-resist.
   
- Breath Blue Magic spell changes:
    - Corrected breath spells to do breath damage, not magic damage.
    - You have to be in front of the monster to do breath damage.
    - REMOVED: Added an angle damage multiplier that lowers damage when monster is more to your side than in front. (requires more info)
    - Damage is now influenced by a monster's elemental resistance, but not by mab/mdb or weather/day bonuses.
    - Damage is now influenced by monster correlation.
    - The resist rate is calculated once and affects both damage and added effect hit rate / duration.
   
- Drain Blue Magic spell changes:
    - Consolidated drain spells into one function.
 
- Curing Blue Magic spell changes:
    - Consolidated curing spells into one function.
 
- General changes:
    - Azure Lore will now allow Physical spells to always skillchain, and Magical spells to always magic burst.
    - Changed all (wrongly named) LUMORIAN and LUMORIAN_KILLER to LUMINIAN and LUMINIAN_KILLER.
    - Added BREATH_DMG_DEALT mod.
    - Updated Mirage Keffiyeh/Mirage Keffiyeh +1/Saurian Helm to have BREATH_DMG_DEALT +10 mods.
    - Added decaying effects to Magic Defense Boost (used if a tick value > 0 is given).

## Steps to test these changes

- Azure Lore
  - Physical spells: try Head Butt (has azuretp value different from multiplier value) under Azure Lore, it should do more damage.
  - Magical spells: try Death Ray (has azureBonus of 2) under Azure Lore, it should do more damage.
  - Try chaining spells together and bursting on top of it - you should be able to continually keep this up as long as Azure Lore is active.

- Physical spells
  - Check accuracy with and without Potency merits, 5/5 Potency gives +10.
  - Chain Affinity: give yourself 3000 TP, Chain Affinity and cast Head Butt. TP should be gone and Head Butt should use the tp300 value (higher multiplier = more damage).
  - Enchainment merits: every merit in Enchainment should give you +100 TP during Chain Affinity. Give yourself 2500 TP and try to see if you get the same result as 3000 TP, CA and no merits.
  - Monster correlation affects multiplier: try Beasts vs Lizards, will have a higher multiplier. Test on spells with a low base multiplier to see a higher effect, should be +.25. Try Beasts vs Plantoids, will be -.25.
  - Added effects: 0 dmg should not result in the AE being applied. Remove your weapon and try landing Head Butt on something - the Stun check should only happen if the spell actually does damage.
  - Decaying effects now work, such as Wild Oats. Can do !getmod VIT to check on a mob.
  - Half-resist on AE should result in half duration. Check magic resist value and check against duration to confirm. Wild Oats is 60s unresisted VIT down, 30s resisted. Any more resist and the AE is not applied.
  - AE all use INT and Blue Magic skill to determine resist rate, check Magic Hit Rate with a ton of these and without.

- Magical spells
  - Try casting spell on 1 vs on multiple mobs - should get a multi-target damage reduction (as described in damage_spell.lua).
  - Monster correlation affects multiplier: try Beasts vs Lizards, will have a higher multiplier. Test on spells with a low base multiplier to see a higher effect, should be +.25. Try Beasts vs Plantoids, will be -.25.
  - Half-resist on AE should result in half duration. Check magic resist value and check against duration to confirm. Blastbomb is 30s unresisted Bind, 15s resisted. Any more resist and the AE is not applied.
  - AE has its own resist check based on Blue Magic skill and the spell's main attribute (INT, MND or CHR), check Magic Hit Rate with a ton of these and without.

- Enhancing spells
  - Decaying effects now work, such as Saline Coat. Can do !getmod VIT Playername to check.

- Enfeebling spells
  - Check that isGaze-flagged spells actually only work when the mob is looking at you and you at them. No effect otherwise.
  - Check that isConal-flagged spells actually only work when the mob is in front of you. No effect otherwise.

- Breath spells
  - Check that Breath spells are not influenced by Mab/Mdb/Weather/Day. (equip Novio earring, no difference. equip Hachirin-no-Obi, go to Uleguerand in Ice weather, cast Frost Breath, remove weather, damage should be no difference).
  - Check that Breath spells are influenced by monster elemental weaknesses, such as Tiamat: Fire -50% (Heat Breath).
  - Check that Breath spells are influenced by level and HP. (change HP and level and see difference).
  - Hit a monster from the front (0°) and more to the side (40°). It should do less damage (45° edge = 50% damage).
  - Monster correlation affects multiplier: try Beasts vs Lizards, should do 25 more damage. Try Beasts vs Plantoids, should be -25%.
  - Mirage Keffiyeh, the +1 and Saurian Helm have a BREATH_DAMAGE_DEALT mod +10. Check that Breath spells do 10% more damage with any of those equipped.

- Cure spells
  - Purely refactored code so we're not repeating ourselves 5 times, but can always check if these still cure.

- Drain spells
  - Check that these return no effect message vs undead.
  - Set a mob's HP to 10 (!hp 10) and use Blood Drain - it should not drain more than 10 HP.
  - Set a mob's MP to 10 (!mp 10) and use MP Drainkiss - it should not drain more than 10 MP.
  - Check if MP Drainkiss only drains MP from mobs that have MP.

- Monster correlation
  - Test that monster correlation merits increase the multiplier by .004 each, 5/5 should be +.02.
  - Test that monster correlation merits only increase positive correlation and do not influence negative correlation.

### CODE REVIEW

**Comment by TeoTwawki** [comment]
> I tested all changes I made.

You still have to describe _how others can._

Give examples, test cases. Steps.

Also serves as another reason to keep you pr's small in size and scope.

**Comment by Enedin9000** [comment]
> > I tested all changes I made.
> 
> You still have to describe _how others can._
> 
> Give examples, test cases. Steps.
> 
> Also serves as another reason to keep you pr's small in size and scope.

Ya, my bad. I added things to test that cover the changes in general, with specific examples to try out.

File: scripts/commands/getstats.lua [hunk 1]
```diff
@@ -49,9 +49,9 @@ function onTrigger(player)
     player:PrintToPlayer(string.format("Magic Attack bonus: %i ", target:getMod(xi.mod.MATT)), xi.msg.channel.SYSTEM_3)
     player:PrintToPlayer(string.format("Magic Defense bonus: %i ", target:getMod(xi.mod.MDEF)), xi.msg.channel.SYSTEM_3)
     player:PrintToPlayer(string.format("Magic Accuracy bonus: %i ", target:getMod(xi.mod.MACC)), xi.msg.channel.SYSTEM_3)
-    -- player:PrintToPlayer(string.format("Total Subtle Blow: %i ", target:getMod(xi.mod.SUBTLE_BLOW)), xi.msg.channel.SYSTEM_3)
-    -- player:PrintToPlayer(string.format("Total Store TP: %i ", target:getMod(xi.mod.STORETP)), xi.msg.channel.SYSTEM_3)
-    -- player:PrintToPlayer("Cannot easily and accurately get Magic Evasion with current methods.")
+    player:PrintToPlayer(string.format("Total Subtle Blow: %i ", target:getMod(xi.mod.SUBTLE_BLOW)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer(string.format("Total Store TP: %i ", target:getMod(xi.mod.STORETP)), xi.msg.channel.SYSTEM_3)
+    player:PrintToPlayer("Cannot easily and accurately get Magic Evasion with current methods.")
```
Comment (Xaver-DaRed) [inline scripts/commands/getstats.lua]: Please revert changes to this file

File: scripts/globals/effects/magic_def_boost.lua [hunk 2]
```diff
@@ -14,10 +14,19 @@ effectObject.onEffectGain = function(target, effect)
 end
 
 effectObject.onEffectTick = function(target, effect)
+    -- decay the effect by 1 every effectTick
+    local boostMDB_effect_size = effect:getPower()
```
Comment (Xaver-DaRed) [inline scripts/globals/effects/magic_def_boost.lua]: This variable could be called just "power"

File: scripts/globals/effects/magic_def_boost.lua [hunk 3]
```diff
@@ -14,10 +14,19 @@ effectObject.onEffectGain = function(target, effect)
...
+    if boostMDB_effect_size > 0 then
+        effect:setPower(boostMDB_effect_size - 1)
+        target:delMod(xi.mod.MDEF, 1)
```
Comment (Xaver-DaRed) [inline scripts/globals/effects/magic_def_boost.lua:21]: if youare setting power to power -1, is it really necesary to also delete 1 point from power?

File: scripts/globals/magic.lua [hunk 4]
```diff
@@ -575,6 +575,7 @@ function getMagicHitRate(caster, target, skillType, element, percentBonus, bonus
     magicacc = magicacc + utils.clamp(maccFood, 0, caster:getMod(xi.mod.FOOD_MACC_CAP))
 
     return calculateMagicHitRate(magicacc, magiceva, percentBonus, caster:getMainLvl(), target:getMainLvl())
+
```
Comment (Xaver-DaRed) [inline scripts/globals/magic.lua]: Unneeded newline

File: scripts/globals/magic.lua [hunk 5]
```diff
@@ -888,12 +889,11 @@ function addBonuses(caster, spell, target, dmg, params)
         elseif casterJob == xi.job.GEO then
             mab = mab + caster:getJobPointLevel(xi.jp.GEO_MAGIC_ATK_BONUS)
         end
-
+
```
Comment (Xaver-DaRed) [inline scripts/globals/magic.lua]: whitespace

File: scripts/globals/mobskills.lua [hunk 6]
```diff
@@ -473,7 +473,8 @@ xi.mobskills.mobBreathMove = function(mob, target, percent, base, element, cap)
     end
 
     -- Deal bonus damage vs mob ecosystem
-    local systemBonus = utils.getSystemStrengthBonus(mob, target)
+    print(string.format("System mob " .. mob:getSystem() .. "/ System target " .. target:getSystem()))
```
Comment (Xaver-DaRed) [inline scripts/globals/mobskills.lua]: Please remove the debug print

File: scripts/globals/spells/blue/amplification.lua [hunk 7]
```diff
@@ -14,8 +14,8 @@
 -- Combos: None
 -----------------------------------
 require("scripts/globals/bluemagic")
+require("scripts/globals/settings")
```
Comment (Xaver-DaRed) [inline scripts/globals/spells/blue/amplification.lua]: settings requirement is unneeded in his file

File: scripts/globals/spells/blue/amplification.lua [hunk 8]
```diff
@@ -28,33 +28,18 @@ spellObject.onSpellCast = function(caster, target, spell)
...
-        target:addStatusEffect(typeEffectTwo, power, 0, duration)
-        spell:setMsg(xi.msg.basic.MAGIC_GAIN_EFFECT)
+    elseif not actionOne and actionTwo then -- the first status fails to apply 
         returnEffect = typeEffectTwo
-    else
-        target:addStatusEffect(typeEffectOne, power, 0, duration)
-        target:addStatusEffect(typeEffectTwo, power, 0, duration)
-        spell:setMsg(xi.msg.basic.MAGIC_GAIN_EFFECT)
+    elseif actionOne and not actionTwo then -- the second status fails to apply 
+        returnEffect = typeEffectOne
```
Comment (Xaver-DaRed) [inline scripts/globals/spells/blue/amplification.lua:41]: Just yesterday I found a message basic that has space to return 2 status effects. This may be a good place to use it.

This is more a note to myself than anything else

File: scripts/globals/effects/magic_def_boost.lua (hunk 3 above)
Comment (Enedin9000) [reply scripts/globals/effects/magic_def_boost.lua:21]: I'm just following the examples on, e.g. VIT down 
image

File: scripts/globals/utils.lua [hunk 9]
```diff
@@ -439,14 +439,11 @@ local systemStrengthTable =
...
-    [xi.eco.LUMINION] = { [xi.eco.LUMORIAN] = 1, },
+    [xi.eco.LUMINIAN] = { [xi.eco.LUMINION] = 1, },
+    [xi.eco.LUMINION] = { [xi.eco.LUMINIAN] = 1, },
 }
 
-function utils.getSystemStrengthBonus(attacker, defender)
-    local attackerSystem = attacker:getSystem()
-    local defenderSystem = defender:getSystem()
-
+function utils.getSystemStrengthBonus(attackerSystem, defenderSystem)
```
Comment (Xaver-DaRed) [inline scripts/globals/utils.lua:446]: Is this function only used by blue spells?

If not, any other system that uses it will still feed it caster and target and fail to work properly

Comment (Enedin9000) [reply scripts/globals/utils.lua:446]: It was only used here, and now by BLU stuff.

File: .vscode/settings.json [hunk 10]
```diff
@@ -6,7 +6,7 @@
 
     "editor.bracketPairColorization.enabled": true,
     "editor.inlayHints.enabled": "offUnlessPressed",
-    "editor.renderWhitespace": "all",
+    "editor.renderWhitespace": "none",
```
Comment (Xaver-DaRed) [inline .vscode/settings.json]: Please revert this change

Comment (Enedin9000) [reply .vscode/settings.json]: bah you'd think "add to gitignore" would do something, but yeah git

File: src/map/modifier.h [hunk 11]
```diff
@@ -955,6 +955,7 @@ enum class Mod
     GEOMANCY_MP_NO_DEPLETE = 1037, // Percent chance for Geomancy to cost 0 MP (GEO AF3 Sets)
 
     DIG_BYPASS_FATIGUE = 1074, // Chocobo digging modifier found in "Blue Race Silks". Modifier works as a direct percent.
+    BREATH_DMG_DEALT = 3033, // Breath damage dealt
```
Comment (Xaver-DaRed) [inline src/map/modifier.h]: Isnt this id... high?
```
    //
    // SPARE IDs:
    // 53
    // 74 to 79
    // 138 to 143
    // 156 to 159
    // 217 to 223
    // 261 to 280
    //
    // SPARE = 1075, and onward
  ```  
    Also, be sure to edit this comments aswell, so we all know which id is used up

Comment (Enedin9000) [reply src/map/modifier.h]: I got it from [1050007342588112958](https://discord.com/channels/392903136336936960/832349765718900827/1050007342588112958)

Comment (Enedin9000) [reply src/map/modifier.h]: no idea how that works tbh

Comment (Xaver-DaRed) [reply src/map/modifier.h]: Simple: Since the IDs for modifiers are made up, you pick the lowest spare id possible

Comment (Enedin9000) [reply src/map/modifier.h]: I've believed in lies

Comment (Xaver-DaRed) [reply src/map/modifier.h]: It's all a dream from Diabolos anyway

File: scripts/globals/status.lua [hunk 12]
```diff
@@ -1818,6 +1818,8 @@ xi.mod =
     BLUE_MAGIC_EFFECT       = 1059, -- TODO: Bonus to Attribute Value of spell (percent)
     QUICK_DRAW_RECAST       = 1060, -- TODO: Quick Draw Charge Reduction (seconds)
 
+    BREATH_DMG_DEALT        = 3033, -- Breath damage dealt
```
Comment (Xaver-DaRed) [inline scripts/globals/status.lua]: you have to change the modifier ID here aswell
They must be the same in both modifier.h and status.lua

File: src/map/utils/blueutils.cpp [hunk 13]
```diff
@@ -44,6 +44,7 @@
 
 namespace blueutils
 {
+
```
Comment (Xaver-DaRed) [inline src/map/utils/blueutils.cpp]: remove this please

File: src/map/status_effect.h [hunk 14]
```diff
@@ -744,12 +744,11 @@ enum EFFECT
     EFFECT_FULL_SPEED_AHEAD    = 803, // Used to track Full Speed Ahead quest minigame
     EFFECT_HYSTERIA            = 804, // Used for Hysteroanima to stop after readying a weaponskill with no msg.
     EFFECT_TOMAHAWK            = 805, // Silent status effect inflicted by a Warrior using the "Tomahawk" job ability
-    // EFFECT_PLACEHOLDER           = 806  // Description
     // 806-1022
     // EFFECT_PLACEHOLDER           = 1023 // The client dat file seems to have only this many "slots", results of exceeding that are untested.
 };
 
-#define MAX_EFFECTID 806 // 768 real + 38 custom
+#define MAX_EFFECTID 806 // 768 real + 39 custom
```
Comment (Xaver-DaRed) [inline src/map/status_effect.h]: no logic changes here, I would say this changes should be removed

File: scripts/globals/effects/magic_def_boost.lua (hunk 3 above)
Comment (Xaver-DaRed) [reply scripts/globals/effects/magic_def_boost.lua:21]: Please, dont resolve conversations, that's my job^^

Also, I want the others to look at this concrete one, so we can know if this is the proper way to do it or if we just found an issue

Comment (Enedin9000) [reply scripts/globals/effects/magic_def_boost.lua:21]: oh shit sorry I thought I was supposed to resolve when I had fixed it

**Comment by Xaver-DaRed** [comment]
Gotta remove that navmesh change from here aswell

**Comment by claywar** [comment]
For navmesh, you might be able to resolve with `git submodule update` and then commit/repush

**Comment by Enedin9000** [comment]
> 

No idea what git submodule update is supposed to do tho but I can just run that command?
Imma see what changed first perhaps I can fix it with what I know
EDIT: Claywar is my hero once again

**Comment by Xaver-DaRed** [comment]
You may or may not need to do it again to remove the wiki update

Unless, that is intended, but I dont remember it being in your original PR

**Comment by Enedin9000** [comment]
> 

euhhh I'm always confused by these things, I dont touch em, I dont want them to change
I did git submodule update --remote again but nothing happens, I assume that means there's nothing to update?

I did it yesterday and both navmeshes and wiki stuff had changes.

**Comment by Xaver-DaRed** [comment]
Overall, It's a very good effort.

Once the wiki submodule issue is dealt with, this should be good to go.

Still, in a future date, most of this logic will be moved and audited in the same manner magical nukes and enfeebling spells have been, but this will certanly makes things easier

**Comment by Enedin9000** [comment]
> Overall, It's a very good effort.
> 
> Once the wiki submodule issue is dealt with, this should be good to go.
> 
> Still, in a future date, most of this logic will be moved and audited in the same manner magical nukes and enfeebling spells have been, but this will certanly makes things easier

I would address the wiki thing but I don't know what to do about it

**Comment by Enedin9000** [comment]
I downloaded the latest documentation from base and put that in my branch and pushed that, idk if that helps we'll see

tbh I'm not sure what the issue is if I could get pointers on how to resolve it that would help. the git command git update submodule --remote isnt doing anything and I never touched any wiki stuff so idk

Like I see changes but can't I just undo all that or somehow ignore that next time? add that to gitignore or something idk 
image

**Comment by Xaver-DaRed** [comment]
Gonna re re review it. Hopefully alls good

File: scripts/globals/utils.lua (hunk 9 above)
Comment (Xaver-DaRed) [reply scripts/globals/utils.lua:446]: So I went ahead and made a mass search

This function is indeed used elsewhere

So please rever this changes, so the function handles the attacker and defender system

Comment (Xaver-DaRed) [reply scripts/globals/utils.lua:446]: EDIT: Ugh, nevermind, i screwed up

File: src/map/entities/battleentity.cpp [hunk 15]
```diff
@@ -1419,7 +1419,7 @@ void CBattleEntity::OnCastFinished(CMagicState& state, action_t& action)
         // TODO: actual radius calculation
         float radius = spell::GetSpellRadius(PSpell, this);
 
-        PAI->TargetFind->findWithinCone(PActionTarget, radius, 45, flags);
+        PAI->TargetFind->findWithinCone(PActionTarget, radius, 90, flags);
```
Comment (Xaver-DaRed) [inline src/map/entities/battleentity.cpp]: Last thing I guess.... What is this change?

Comment (Enedin9000) [reply src/map/entities/battleentity.cpp]: This is used in player magic conal spells. the only such ones I could find were all BLU spells but a 45° cone is a bit small from what I remember on retail so I increased it to 90°. This gives Breath spells some room to have some noticeable damage modulation when the monster is not as much in front, as it was on retail. This pairs with Breath stuff in bluemagic.lua:
image
Of course this also works with a smaller cone it's just less visible.

Comment (Xaver-DaRed) [reply src/map/entities/battleentity.cpp]: that would be a 45 / 256 irrc. Not a 45 / 360. Or not, I may be wrong.

Still 90 degrees / 360 seems excesive aswell. for now, I will request this change to be reverted until we can actually verify it in retail

Comment (Enedin9000) [reply src/map/entities/battleentity.cpp]: ok, I'll put it back to 45 and restrict the angle in Breath stuff in bluemagic.lua as well

**Review by Xaver-DaRed — APPROVED** [review]

**Comment by PrettzL** [comment]
Noting that on retail you do not need to be in front of a mob, let alone facing it, to deal full breath damage to it: https://youtu.be/_KqAEsdsLZo?t=277

**Comment by claywar** [comment]
> Noting that on retail you do not need to be in front of a mob, let alone facing it, to deal full breath damage to it: https://youtu.be/_KqAEsdsLZo?t=277

@PrettzL Since this PR has been merged, if there are any issues regarding it, I would suggest raising a bug so that it can be addressed in the future.

**Comment by Enedin9000** [comment]
> Noting that on retail you do not need to be in front of a mob, let alone facing it, to deal full breath damage to it: https://youtu.be/_KqAEsdsLZo?t=277

Interesting, you'd think Breath would require you to face the target. The code doesn't require the mob to look at you for Breaths, that's for Gaze moves.
